### PR TITLE
Homework10.2

### PR DESCRIPTION
**Pull request by jfdsmit**

### CODE REVIEW

File: course5/homework10.2/linked_list.h [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+namespace jds {
+
+template <typename T>
+class LinkedList
+{
+private:
+	struct Node
+	{
+		T value;
+		std::unique_ptr<Node> next{nullptr};
```
Comment (maciekgajewski) [inline course5/homework10.2/linked_list.h]: nullptr not needed

File: course5/homework10.2/linked_list.h [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+		const_iterator & operator++() {
+			current = current->next.get();
+			return *this;
+		}
+		const_iterator operator++(int) {
+			const_iterator old = *this;
+			current = current->next.get();
+			return old;
+		}
+		T const & operator*() const { return current->value; }
```
Comment (maciekgajewski) [inline course5/homework10.2/linked_list.h:66]: not sure about this..
Should you have
```
T& operator*();
const T& operator*() const;
```
?

File: course5/homework10.2/linked_list.h [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+		bool operator==(const_iterator const & that) const { return this->current == that.current; }
+		bool operator!=(const_iterator const & that) const { return this->current != that.current; }
+	private:
+		Node const * current{nullptr};
+	};
```
Comment (maciekgajewski) [inline course5/homework10.2/linked_list.h:72]: ```operator->``` is missing

File: course5/homework10.2/linked_list.h [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+	LinkedList() = default;
+	LinkedList(LinkedList const & that): LinkedList(that.begin(), that.end()) { }
+	LinkedList(LinkedList && that): head(std::move(that.head)) { }
+
+	template <class InputIt>
+	LinkedList(InputIt const & first, InputIt const & last)
+	{
+		std::unique_ptr<Node> * current{&head};
```
Comment (maciekgajewski) [inline course5/homework10.2/linked_list.h]: Why are you using this curly-bracket initialisation everywhere? Makes code harder to read.

File: course5/homework10.2/linked_list.h [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+			*current = std::make_unique<Node>();
+			(*current)->value = *it;
+			current = &((*current)->next);
+		}
+	}
+
+	LinkedList & operator=(LinkedList const & that)
+	{
+		this->head = nullptr;
+		std::unique_ptr<Node> * current{&head};
```
Comment (maciekgajewski) [inline course5/homework10.2/linked_list.h]: This is the same non-trivial code as in the function above. Consider putting it into common function.

File: course5/homework10.2/linked_list.h [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+			current = &((*current)->next);
+		}
+		return *this;
+	}
+		
+	LinkedList & operator=(LinkedList && that) { this->head = std::move(that.head); return *this; }
+
+	bool operator==(LinkedList const & that) const
+	{
+		auto itl = begin(), itr = that.begin();
```
Comment (maciekgajewski) [inline course5/homework10.2/linked_list.h]: Consider using ```std::equal```

File: course5/homework10.2/linked_list.h [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+	template <class... Args>
+	void emplace_front(Args&& ... args)
+	{
+		std::unique_ptr<LinkedList<T>::Node> newNode = std::make_unique<LinkedList<T>::Node>();
+		newNode->value = T(std::forward<Args>(args)...);
+		newNode->next = std::move(head);
+		head = std::move(newNode);
+	}
+
+	void pop_front() { if (head) head = std::move(head->next); }
```
Comment (maciekgajewski) [inline course5/homework10.2/linked_list.h:118]: The ```if``` breaks "don't pay for what you're not using" principle.

File: course5/homework10.2/linked_list.h [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+	T & front() const { return head->value; }
+	std::size_t size()
+	{
+		std::size_t result{0};
+		for (auto it = begin(); it != end(); ++it)
```
Comment (maciekgajewski) [inline course5/homework10.2/linked_list.h]: Consider using ```std::distance```

**Review by maciekgajewski — COMMENTED** [review]
Nice!
But your list has one flaw: it requires T to be default-constructible.

Try creating a type without default constructor, and storing it in your list.

**Comment by jfdsmit** [comment]
Yeah, I realized the default construction problem on the train, tried to fix it on the train.

File: course5/homework10.2/linked_list.h (hunk 7 above)
Comment (jfdsmit) [reply course5/homework10.2/linked_list.h:118]: What's the best solution? Document that popping from an empty list is undefined and let the program crash?

**Pull request closed without merging**